File: expose-countries-java/src/main/java/br/com/edumazaro/exposecountries/service/RegionService.java
package br.com.edumazaro.exposecountries.service;

import br.com.edumazaro.exposecountries.model.entity.Region;
import br.com.edumazaro.exposecountries.repositories.RegionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class RegionService {

    private RegionRepository regionRepository;

    public RegionService(RegionRepository regionRepository) {
        this.regionRepository = regionRepository;
    }

    public List<Region> findAll() {
        return regionRepository.findAll();
    }

    public Page<Region> findByRegionAndKeyWord(String region, Integer keyWord, Pageable pageable) {
        return regionRepository.findByRegionAndKeyWord(region, keyWord, pageable);
    }
}

File: expose-countries-go/application/region_service.go
package application

import "expose-countries-go/domain"

type RegionService struct {
	repo domain.RegionRepository
}

func NewRegionService(repo domain.RegionRepository) *RegionService {
	return &RegionService{repo: repo}
}

func (service *RegionService) GetRegionsByCriteria(region string, keyWord int, page, perPage int) ([]domain.Region, int64, error) {
	offset := (page - 1) * perPage
	return service.repo.FindByRegionAndKeyWord(region, keyWord, offset, perPage)
}
